perception/detected_object_validation/src/obstacle_pointcloud_based_validator.cpp

```diff
@@ -85,24 +85,109 @@ ObstaclePointCloudBasedValidator::ObstaclePointCloudBasedValidator(
+    return;
+  }
+
+  // Convert to PCL
+  pcl::PointCloud<pcl::PointXYZ>::Ptr obstacle_pointcloud(new pcl::PointCloud<pcl::PointXYZ>);
+  pcl::fromROSMsg(*input_obstacle_pointcloud, *obstacle_pointcloud);
+  if (obstacle_pointcloud->empty()) {
+    return;
+  }
+
+  // Create Kd-tree to search neighbor pointcloud to reduce cost
+  pcl::search::Search<pcl::PointXYZ>::Ptr kdtree =
+    pcl::make_shared<pcl::search::KdTree<pcl::PointXYZ>>(false);
+  kdtree->setInputCloud(obstacle_pointcloud);
+  for (size_t i = 0; i < transformed_objects.objects.size(); ++i) {
+    const auto & transformed_object = transformed_objects.objects.at(i);
+    const auto object_label_id = transformed_object.classification.front().label;
+    const auto & object = input_objects->objects.at(i);
+    const auto & transformed_object_position =
+      transformed_object.kinematics.pose_with_covariance.pose.position;
+    const auto search_radius = getMaxRadius3D(transformed_object);
+    if (!search_radius) {
+      output.objects.push_back(object);
+      continue;
+    }
+    // Search neighbor pointcloud to reduce cost
+    pcl::PointCloud<pcl::PointXYZ>::Ptr neighbor_pointcloud(new pcl::PointCloud<pcl::PointXYZ>);
+    std::vector<int> indices;
+    std::vector<float> distances;
+    pcl::PointXYZ trans_obj_position_pcl;
+    trans_obj_position_pcl.x = transformed_object_position.x;
+    trans_obj_position_pcl.y = transformed_object_position.y;
+    trans_obj_position_pcl.z = transformed_object_position.z;
+    kdtree->radiusSearch(trans_obj_position_pcl, search_radius.value(), indices, distances);
+    for (const auto & index : indices) {
+      neighbor_pointcloud->push_back(obstacle_pointcloud->at(index));
+    }
+
+    if (debugger_) debugger_->addNeighborPointcloud(neighbor_pointcloud);
+    // Filter object that have few pointcloud in them
+    const auto num = getPointCloudNumWithinShape(transformed_object, neighbor_pointcloud);
+    const auto object_distance =
+      std::hypot(transformed_object_position.x, transformed_object_position.y);
+    size_t min_pointcloud_num = std::clamp(
+      static_cast<size_t>(
+        points_num_threshold_param_.min_points_and_distance_ratio.at(object_label_id) /
+          object_distance +
+        0.5f),
+      static_cast<size_t>(points_num_threshold_param_.min_points_num.at(object_label_id)),
+      static_cast<size_t>(points_num_threshold_param_.max_points_num.at(object_label_id)));
+    if (num) {
+      (min_pointcloud_num <= num.value()) ? output.objects.push_back(object)
+                                          : removed_objects.objects.push_back(object);
+    } else {
+      output.objects.push_back(object);
+    }
+  }
+  objects_pub_->publish(output);
+  if (debugger_) {
+    debugger_->publishRemovedObjects(removed_objects);
+    debugger_->publishNeighborPointcloud(input_obstacle_pointcloud->header);
+    debugger_->publishPointcloudWithinPolygon(input_obstacle_pointcloud->header);
+  }
+}
```

Comment:
Since most of them are the same as callback functions of XY-plane, callback functions may be common.

Reply:
@yukkysaito thank you for your comment. I changed  at https://github.com/autowarefoundation/autoware.universe/pull/5327/commits/e3cbf6cb524a34ad3b693a8c0eaa1680e3141f72

Reply:
I'm sorry, I didn't communicate it well.
It was a little different from my image.

Most parts of the callback function are the same for 2D and 3D, so I would be happy if you could abstract and absorb only the parts that make a difference in the interface (using template, function pointer, etc.). If it is difficult, it might be better to have separate callbacks.

Reply:
@yukkysaito I am sorry for late response. Currently, I did not find the efficient way to refactor as your comment yet, so I would like revert to the separate callbacks way first. I will consider to refactor this ASAP in another PR.